client/exec.go

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/canonical/pebble/internal/wsutil"
+)
+
+// ExecOptions are the main options for the Exec call.
+type ExecOptions struct {
+	// Required: command and arguments (first element is the executable)
+	Command []string
+
+	// Optional environment variables
+	Environment map[string]string
+
+	// Optional working directory (default is $HOME or "/" if $HOME not set).
+	WorkingDir string
+
+	// Optional timeout for the command execution, after which the process
+	// will be terminated. If zero, no timeout applies.
+	Timeout time.Duration
+
+	// Optional user ID and group ID for the process to run as.
+	UserID  *int
+	User    string
+	GroupID *int
+	Group   string
+
+	// True to ask the server to set up a pseudo-terminal (PTY) -- this allows
+	// full interactivity and window resizing. The default is no PTY, and just
+	// to use pipes for stdin/stdout/stderr.
+	Terminal bool
+
+	// True to separate the process's stderr into a separate websocket. The
+	// default is to send combined stdout+stderr on a single websocket.
+	//
+	// Note: currently only the combinations Terminal=true, Stderr=false and
+	// Terminal=false, Stderr=true are supported.
+	Stderr bool
+
+	// Initial terminal width and height (only apply if Terminal is true).
+	// If not specified, the Pebble server uses the target's default (usually
+	// 80x25). When using the "pebble exec" CLI, these are set to the host's
+	// terminal size automatically.
+	Width  int
+	Height int
+}
+
+// ExecAdditionalArgs are additional control arguments for the Exec call.
+type ExecAdditionalArgs struct {
```

Comment:
Yeah, fair enough. The short answer is "because LXD separated them", I guess because some go directly to the API as parameters, and some are more for the Go client, but I don't think it makes much sense. The Python client has them all as kwargs, so a single struct seems better here. Fixed, and renamed `Stderr bool` to `SeparateStderr bool` (to avoid the conflict with the `Stderr` stream) and renamed `Terminal bool` to `UseTerminal bool`, both of which are more explicit anyway.